docs/strategies/hedging.md

```diff
@@ -73,6 +73,133 @@ You can use the following special values for `Delay` or in `DelayGenerator`:
 - `0 seconds` - the hedging strategy immediately creates a total of `MaxHedgedAttempts` and completes when the fastest acceptable result is available.
 - `-1 millisecond` - this value indicates that the strategy does not create a new hedged task before the previous one completes. This enables scenarios where having multiple concurrent hedged tasks can cause side effects.
 
+## Diagrams
+
+### Happy path sequence diagram
+
+The hedging strategy does not trigger because the response arrives faster than the threshold.
+
+```mermaid
+sequenceDiagram
+    autonumber
+    actor C as Caller
+    participant P as Pipeline
+    participant H as Hedging
+    participant D as DecoratedUserCallback
+    participant LB as LoadBalancer
+    participant S1 as ServiceReplica1
+    participant S2 as ServiceReplica2
+
+    C->>P: Calls ExecuteAsync
+    P->>H: Calls ExecuteCore
+    Note over H: Wait start
+    activate H
+
+    H->>+D: Invokes
+    D->>LB: Issues request
+    LB->>+S1: Forwards request
+    S1-->>S1: Processes request
+    S1->>-D: Returns result
+    D->>-H: Returns result
+    deactivate H
+    Note over H: Wait end
+    H->>P: Returns result
+    P->>C: Returns result
+```
+
+### Unhappy path sequence diagram
+
+The hedging strategy triggers because the response arrives slower than the threshold.
+
+```mermaid
+sequenceDiagram
+    autonumber
+    actor C as Caller
+    participant P as Pipeline
+    participant H as Hedging
+    participant D as DecoratedUserCallback
+    participant LB as LoadBalancer
+    participant S1 as ServiceReplica1
+    participant S2 as ServiceReplica2
+
+    C->>P: Calls ExecuteAsync
+    P->>H: Calls ExecuteCore
+    Note over H: Wait start
+    activate H
+    H->>D: Invokes
+    D->>LB: Issues request
+    LB->>S1: Forwards request (R1)
+    activate S1
+    S1-->>S1: Processes request<br/> slowly ...
+
+    Note over H: Wait end
+    H->>LB: Issues request
+    LB->>+S2: Forwards request (R2)
+    S2->>S2: Processes request
+    S1-->>S1: Still processing ...
+    S2->>-D: Returns result (R2)
+    D->>H: Returns result (R2)
+    H->>S1: Propagates Cancellation
+    deactivate S1
+    deactivate H
+
+    H->>P: Returns result (R2)
+    P->>C: Returns result (R2)
+```
+
+### Alternative: happy path sequence diagram
+
+```mermaid
+sequenceDiagram
+    autonumber
+    actor C as Caller
+    participant P as Pipeline
+    participant H as Hedging
+    participant HUC as HedgedUserCallback
+
+    C->>P: Calls ExecuteAsync
+    P->>H: Calls ExecuteCore
+    Note over H: Wait start
+    activate H
+    H->>+HUC: Invokes
+    HUC-->>HUC: Processes request
+    HUC->>-H: Returns result
+    deactivate H
+    Note over H: Wait end
+    H->>P: Returns result
+    P->>C: Returns result
+```
+
+### Alternative: Unhappy path sequence diagram
+
+```mermaid
+sequenceDiagram
+    autonumber
+    actor C as Caller
+    participant P as Pipeline
+    participant H as Hedging
+    participant D as HedgedUserCallback
```

Comment:
participant C as UserCallback
participant G as ActionGenerator

- The first request executes UserCallback directly
- Second and all other create `HedgedUserCallback` from `ActionGenerator`

Reply:
I think omitting the `ActionGenerator` from the majority of the diagrams eases the understanding.

We could add a single diagram to the *Action Generator* which shows the usage of the generator.

Can you help me by creating that diagram?

Reply:
Okay, I will add that diagram into the next PR.